Feature/tao 7606/search eligibility api

ks16 opened the pull request:
Create API to search for eligibility.
Improve error messages in Eligibility API when one of required parameters is missing.

Commit:
Create REST API controller to search for eligibility.

Commit:
Update API documentation in swagger.json

Commit:
Bump version

Commit:
Return clear message with name of missing required parameter

peetya commented on controller/AbstractRestController.php:
```diff
@@ -44,10 +44,12 @@
-        $testCenterUri = $this->getParameterFromRequest(self::PARAMETER_TEST_CENTER_ID);
-
         try {
+            $testCenterUri = $this->getParameterFromRequest(self::PARAMETER_TEST_CENTER_ID);
```
As the variable is used in the exception, please initialize it outside in the try `try` scope.

peetya commented on controller/RestEligibilities.php:
```diff
@@ -0,0 +1,144 @@
+                'data' => $result
+            ]);
+        } catch (\Exception $e) {
+            return $this->returnFailure($e);
```
You can leave the `return` as this method won't return anything.

peetya commented on controller/RestEligibilities.php:
```diff
@@ -0,0 +1,144 @@
+
+    /**
+     * @OA\Get(
+     *     path="/taoTestCenter/api/eligibilities",
```
I think it's not straightforward that we have and endpoint which indicates that the result will be multiple resources, but it can return only one resource. I would say this logic should be done in `/taoTestCenter/api/eligibility`. What do you think?

ks16 replied:
We can not easily implement it in this endpoint because all GET requests are routed to get() method in controller and it's already used to get eligibility by id. I will return array of resources here. I future we need to improve this endpoint to perform real search by one of parameters with pagination and etc.

peetya replied:
It's better now :) Could you please update the the swagger documentations as well?

peetya commented on controller/RestEligibilities.php:
```diff
@@ -0,0 +1,144 @@
+        } catch (common_exception_MissingParameter $e) {
+            throw new common_exception_RestApi(__('Missed required parameter: `%s`', self::PARAMETER_DELIVERY_ID));
+        } catch (common_exception_NotFound $e) {
+            throw new common_exception_RestApi("Delivery `{$deliveryUri}` does not exist.");
```
Please initialize the `$deliveryUri` outside the `try-catch` block.

peetya commented on controller/RestEligibility.php:
```diff
@@ -379,10 +379,12 @@ private function getEligibilityFromRequest()
-        $deliveryUri = $this->getParameterFromRequest(self::PARAMETER_DELIVERY_ID);
-
         try {
+            $deliveryUri = $this->getParameterFromRequest(self::PARAMETER_DELIVERY_ID);
```
Same problem.

peetya commented on doc/swagger.json:
```diff
@@ -5,6 +5,71 @@
+              "application/json": {
+                "schema": {
+                  "example": {
+                    "delivery": "http://sample/first.rdf#i1536680377163170",
```
The `success` attribute is missing from the response.

peetya commented on controller/RestEligibilities.php:
```diff
@@ -0,0 +1,144 @@
+     * @return \core_kernel_classes_Resource
+     * @throws common_exception_RestApi
+     */
+    private function getDeliveryFromRequest()
```
This function is already exist in `RestEligibility` class. Would be better to move it into the abstract class. What do you think?

ks16 replied:
This abstract class is common for few other controllers as well, which do not need delivery. Therefore I wouldn't move it. But it's just my personal opinion

peetya commented on doc/swagger.json:
```diff
@@ -5,6 +5,71 @@
+              }
+            }
+          },
+          "404": {
```
`400` errors are missing as it could return this error in case of non existing deliveries or test-centers.

peetya commented on controller/RestEligibilities.php:
```diff
@@ -0,0 +1,144 @@
+     *         ),
+     *     ),
+     *     @OA\Response(
+     *         response=404,
```
It will never return 404 as the empty response is handled in `RestEligibilites::get()` in line 111.

Commit:
Return list of resources.
Initiate uri variables outside of try block to ensure it exists in catch block.

Commit:
Remove unnecessary return statement

Commit:
Update swagger annotations and API docummentation in swagger.json.
